doc/man/e2-create-project.1
.\" Man page for e2-create-project
.\"
.\" (c)2007 emlix GmbH
.\"
.TH e2-create-project 1 "Aug 6, 2007" "0.1"

.SH NAME
e2-create-project \- create a new project

.SH SYNOPSIS
e2-create-project [<option> ...] [<server>:]<location>

.SH DESCRIPTION
Create a new project and store it on <server> in <location>.
<server> defaults to 'projects'.
The project name will be the basename of <location>.
The usual project location convention is
\fBprojects:<customer>/<projectname>\fR.

.SH RETURN VALUE
Normally, exit status is 0. On error, it is non-zero.

.SH OPTIONS
.TP
.BR \-\-help
Show usage information.
.TP
.BR \-\-version
Show version number of e2.

.P
For further global options, see \fBe2intro\fR(7).

.SH BUGS
Doubtless there are many.

.SH "SEE ALSO"
.BR e2intro(7)
.BR e2global(7)
.BR e2-fetch-project(1)
.BR e2-fetch-sources(1)
.BR e2-ls-project(1)
.BR e2config(5)

.SH AUTHORS
Gordon Hecker, Oskar Schirmer and Felix Winkelmann

.SH COPYRIGHT
(c)2007 emlix GmbH


Changelog
NEXT:

e2factory-2.3.12
 * allow server paths to contain whitespace
 * bugfixes

e2factory-2.3.11
 * remove bashism, noted by Dominic Lemire
 * release to sync with extensions

e2factory-2.3.10
 * add method for plugins to register additional tools

e2factory-2.3.9
 * e2-build: allow enabling/disabling writeback on command line
 * use rsync to create remote directories when using rsync+ssh url, to
   simplify restricted-shell setups

e2factory-2.3.8
 * Bugfix: Large files (>2GB) in cache are detected correctly now on 32bit
 * Bugfix: e2-install-e2 reports errors in local tools installation now
 * Bugfix: e2-install-e2 removes an existing local plugin directory before
   starting local tools installation
 * Minor Bugfixes

e2factory-2.3.7
 * BuildIds are now calculated in advance to detect errors immediately
   before starting the first build process
 * remove old local installation before re-installing. Plugins were not
   removed in e2-install-e2 before.

e2factory-2.3.6
 * the release build process changed: Results are now stored on the
   results server, results listed in the deploy_results list are
   deployed to the releases server for archive purposes.
 * minor bugfixes
 * rsync is now used instead of curl to access local files
 * allow environment configuration in source and results configuration
   files

e2factory-2.3.5
 * bugfixes in collect_project: Copying files was not handled correctly

e2factory-2.3.4
 * version information for plugins is displayed now when requested
 * configuration file setting in .e2/e2config is honoured in e2-install-e2
   now
 * bugs and documentation issues in e2-new-source were fixed
 * the plugin interface was extended to permit plugins to register
   commandline options and permit tool specific initialization of plugins
 * a serious bug in the hashcache was fixed. The buildid calculation
   used outdated hash of files from the project tree.
 * the username is included in the chroot environment path again
 * the plugin interface was extended to allow plugins to change the
   way dependencies are processed into the build order
 * bugs were fixed
 * compressed build log is stored within the result packages
 * bugs were fixed
 * a plugin interface was introduced. Local code is capable of loading
   plugins and some initial interfaces are available with this release.
 * the environment related code was rewritten in a far more robust way.
   This fixes a few bugs: Missing environment files were silently ignored,
   and adding result specific environment could change the id of the
   global environment (not the environment itself) due to lua table behaviour.
 * the cache module was turned into a true lua module.
 * a design flaw in the cache module was fixed that lead to incorrect
   caching for the builtin proj-storage server. Caching was disabled due
   to this problem in the default configuration and is enabled now.
 * support for per-project caches (using the %l format) in e2.conf is
   no longer available. Cache is designed to be per user.
 * the global interface version is no longer maintained in the file
   .e2/version inside the project. Instead ./version on the server side
   remains the only permanent copy and it is fetched to .e2/global-version
   in the project environment
 * the local tools check for their own support for the global interface
   version in use.
 * a few minor bugs were fixed
 * code related to external tools was moved into the new tools module
 * a lot of effort was done to further modularize code
 * fix alignment of status messages for better readability
 * tool invocations are no longer logged

e2factory-2.3.3
 * use sha1 module instead of calling the sha1sum tool
 * optimization: use stat() instead of calling 'test -f'
 * rotate build logfiles
 * introduce a hash cache depending on timestamps to further speed up
   BuildId calculation
 * release mode verifies hashes for server side files
 * release mode checks that the correct version of e2factory is used with
   a project

e2factory-2.3.2
 * temporary build directories are located in a version specific directory
   now
 * e2-install-e2 installs branches properly now. Git refs can be specified.
 * a bug in e2-build --force-rebuild was fixed that prevented forced rebuilds
 * the chroot environment is locked now when in use
 * hardlinks are used when preparing the result tarball, if possible, to
   improve performance
 * logfiles are rotated instead of deleting them
 * a bug in configuration validation for files sources was fixed
 * a bug in validating the licences attribute in source configurations was
   fixed
 * the location of a patch that does not apply is included in the error
   message now
 * unused code was removed
 * the faq section in the documentation was extended
 * outdated documentation was removed
 * a bug was fixed that lead to partial results being left on local servers
   or in the cache when copying was interrupted
 * storing results to cache and to servers is optimized by trying to
   create hardlinks if possible. This may fail due to filesystem boundaries.
   In that case the code falls back to copying.
   Hardlinking is not possible if push_permissions is set for the particular
   server.
 * the configuration file format for chroot and licence configuration files
   has changed to prepare for configuring per-file checksums where files
   are referenced
 * sha1 checksums must be configured in the source-, licence- and chroot
   configuration now where files are referenced. Exceptions are allowed
   for files stored in the project configuration tree (on the "." server)
 * fix the bug that the project name configuratino from `proj/config` did
   not influence the BuildId
 * the example project is now included in the main source tree. Additional
   files and the e2factory.git repository can be fetched on-the-fly

e2factory-2.3.1
 * bugfixes

e2factory-2.3.1rc1
 * bugfixes, documentation

e2factory-2.3.1pre1
 * bugfixes

global/e2-create-project.lua
--[[
   e2factory, the emlix embedded build system

   Copyright (C) 2007-2009 Gordon Hecker <gh@emlix.com>, emlix GmbH
   Copyright (C) 2007-2009 Oskar Schirmer <os@emlix.com>, emlix GmbH
   Copyright (C) 2007-2008 Felix Winkelmann, emlix GmbH

   For more information have a look at http://www.e2factory.org

   e2factory is a registered trademark by emlix GmbH.

   This file is part of e2factory, the emlix embedded build system.

   e2factory is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.
]]

require("buildconfig")
require("e2lib")
require("e2option")
local cache = require("cache")
local generic_git = require("generic_git")

e2lib.init()

local doc = [[
usage: e2-create-project [<option> ...] [<server>:]<location>

Create a new project and store it on <server> in <location>.
<server> defaults to '%s'.
]]
e2option.documentation = string.format(doc,
				e2lib.globals.default_projects_server)

local opts = e2option.parse(arg)
local rc, e = e2lib.read_global_config()
if not rc then
  e2lib.abort(e)
end
e2lib.init2()
local e = new_error("creating project failed")

local config, re = e2lib.get_global_config()
if not config then
  e2lib.abort(e:cat(re))
end
local scache, re = e2lib.setup_cache()
if not scache then
  e2lib.abort(e:cat(re))
end

-- standard global tool setup finished

if e2lib.globals.osenv["E2_LOCAL_TAG"] and e2lib.globals.osenv["E2_LOCAL_BRANCH"] then
  e2lib.globals.local_e2_branch = e2lib.globals.osenv["E2_LOCAL_BRANCH"]
  e2lib.globals.local_e2_tag = e2lib.globals.osenv["E2_LOCAL_TAG"]
elseif e2lib.globals.osenv["E2_LOCAL_TAG"] then
  e2lib.globals.local_e2_branch = "-"
  e2lib.globals.local_e2_tag = e2lib.globals.osenv["E2_LOCAL_TAG"]
elseif e2lib.globals.osenv["E2_LOCAL_BRANCH"] then
  e2lib.globals.local_e2_branch = e2lib.globals.osenv["E2_LOCAL_BRANCH"]
  e2lib.globals.local_e2_tag = "^"
else
  e2lib.globals.local_e2_branch = config.site.e2_branch
  e2lib.globals.local_e2_tag =  config.site.e2_tag
end

if #opts.arguments ~= 1 then
  e2option.usage(1)
end

local sl, re = e2lib.parse_server_location(opts.arguments[1],
					e2lib.globals.default_projects_server)
if not sl then
  e2lib.abort(e:cat(re))
end

local p = {}
p.version = buildconfig.GLOBAL_INTERFACE_VERSION[1] -- the project version
p.e2version = string.format("%s %s", e2lib.globals.local_e2_branch,
						e2lib.globals.local_e2_tag)
p.server = sl.server				-- the server
p.location = sl.location			-- the project location
p.name = e2lib.basename(sl.location)		-- the project basename
p.server = sl.server				-- the server

-- create the server side structure
local tmpdir = e2lib.mktempdir()
e2lib.chdir(tmpdir)

local version = string.format("%d\n", p.version)
local empty = ""
files = {
	{ filename = "version", content=version },
	{ filename = "proj/.keep", content=empty },
	{ filename = "git/.keep", content=empty },
	{ filename = "files/.keep", content=empty },
	{ filename = "cvs/.keep", content=empty },
	{ filename = "svn/.keep", content=empty },
}
for _,f in ipairs(files) do
	local dir = e2lib.dirname(f.filename)
	rc, re = e2lib.mkdir(dir, "-p")
	if not rc then
		e2lib.abort(e:cat(re))
	end
	rc, re = e2lib.write_file(f.filename, f.content)
	if not rc then
		e2lib.abort(e:cat(re))
	end
	local sourcefile = string.format("%s/%s", tmpdir, f.filename)
	local flocation = string.format("%s/%s", p.location, f.filename)
	local cache_flags = {}
	rc, re = cache.push_file(scache, sourcefile, p.server, flocation,
								cache_flags)
	if not rc then
		e2lib.abort(e:cat(re))
	end
end
e2lib.chdir("/")
e2lib.rmtempdir(tmpdir)

local tmpdir = e2lib.mktempdir()
e2lib.chdir(tmpdir)

-- create the initial repository on server side
local rlocation = string.format("%s/proj/%s.git", p.location, p.name)
local rc, re = generic_git.git_init_db(scache, p.server, rlocation)
if not rc then
	e2lib.abort(e:cat(re))
end

-- works up to this point

-- create the initial (git) repository
local url = string.format("file://%s/.git", tmpdir)
rc, re = e2lib.git(nil, "init-db")
if not rc then
	e2lib.abort(e:cat(re))
end

local gitignore = e2lib.read_template("gitignore")
if not gitignore then
  e2lib.abort(re)
end
local chroot, re = e2lib.read_template("proj/chroot")
if not chroot then
  e2lib.abort(re)
end
local licences, re = e2lib.read_template("proj/licences")
if not licences then
  e2lib.abort(re)
end
local env, re = e2lib.read_template("proj/env")
if not env then
  e2lib.abort(re)
end
local pconfig, re = e2lib.read_template("proj/config")
if not pconfig then
  e2lib.abort(re)
end
pconfig = pconfig:gsub("<<release_id>>", p.name)
pconfig = pconfig:gsub("<<name>>", p.name)
local name = string.format("%s\n", p.name)
local release_id = string.format("%s\n", p.name) -- use the name for now
local version = string.format("%s\n", p.version)
local e2version = string.format("%s\n", p.e2version)
local syntax = string.format("%s\n", buildconfig.SYNTAX[1])
local empty = ""
files = {
	{ filename = ".e2/.keep", content=empty },
	{ filename = "in/.keep", content=empty },
	{ filename = "log/.keep", content=empty },
	{ filename = "proj/init/.keep", content=empty },
	{ filename = "res/.keep", content=empty },
	{ filename = "src/.keep", content=empty },
	{ filename = "proj/chroot", content=chroot },
	{ filename = "proj/licences", content=licences },
	{ filename = "proj/env", content=env },
	{ filename = "proj/config", content=pconfig },
	{ filename = ".e2/syntax", content=syntax },
	{ filename = ".e2/e2version", content=e2version },
	{ filename = ".gitignore", content=gitignore },
}
for _,f in ipairs(files) do
	local dir = e2lib.dirname(f.filename)
	rc, re = e2lib.mkdir(dir, "-p")
	if not rc then
		e2lib.abort(e:cat(re))
	end
	rc, re = e2lib.write_file(f.filename, f.content)
	if not rc then
		e2lib.abort(e:cat(re))
	end
	rc, re = e2lib.git(nil, "add", f.filename)
	if not rc then
		e2lib.abort(e:cat(re))
	end
end
rc, re = e2lib.write_extension_config(config.site.default_extensions)
if not rc then
	e2lib.abort(e:cat(re))
end
rc, re = e2lib.git(nil, "add", e2lib.globals.extension_config)
if not rc then
	e2lib.abort(e:cat(re))
end
rc, re = e2lib.git(nil, "commit", "-m \"project setup\"")
if not rc then
	e2lib.abort(e:cat(re))
end

local refspec = "master:refs/heads/master"
local rlocation = string.format("%s/proj/%s.git", p.location, p.name)
rc, re = generic_git.git_push(scache, ".git", p.server, rlocation, refspec)
if not rc then
	e2lib.abort(e:cat(re))
end

e2lib.chdir("/")
e2lib.rmtempdir(tmpdir)
e2lib.finish()
